File: tests/src/test/java/com/optivem/atddaccelerator/templategenerator/systemtest/ScriptTest.java
package com.optivem.atddaccelerator.templategenerator.systemtest;

import com.optivem.atddaccelerator.templategenerator.systemtest.clients.GithubClient;
import com.optivem.atddaccelerator.templategenerator.systemtest.clients.GeneratorClient;
import com.optivem.atddaccelerator.templategenerator.systemtest.dsl.GeneratorDsl;
import com.optivem.atddaccelerator.templategenerator.systemtest.dsl.GitHubDsl;
import com.optivem.atddaccelerator.templategenerator.systemtest.util.Constants;
import com.optivem.atddaccelerator.templategenerator.systemtest.util.Language;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Disabled;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.Arguments;
import org.junit.jupiter.params.provider.MethodSource;

import java.util.stream.Stream;

import static org.assertj.core.api.Assertions.assertThat;

class ScriptTest {


    private static final String REPO_OWNER = "valentinajemuovic";

    private GeneratorDsl generator;
    private GitHubDsl gitHub;
    private String repoName;

    @BeforeEach
    void setup() {
        repoName = newName();
        generator = new GeneratorDsl(new GeneratorClient());
        gitHub = new GitHubDsl(new GithubClient(REPO_OWNER, repoName));
    }

    @AfterEach
    void teardown() {
        var created = generator.isCreated(repoName);

        if(created && gitHub != null) {
            gitHub.deleteRepository();
        }
    }

    static Stream<Arguments> languageProvider() {
        return Stream.of(
                Arguments.of(Language.DOTNET, Language.DOTNET),
                Arguments.of(Language.DOTNET, Language.JAVA),
                Arguments.of(Language.DOTNET, Language.TYPESCRIPT),

                Arguments.of(Language.JAVA, Language.DOTNET),
                Arguments.of(Language.JAVA, Language.JAVA),
                Arguments.of(Language.JAVA, Language.TYPESCRIPT),

                Arguments.of(Language.TYPESCRIPT, Language.DOTNET),
                Arguments.of(Language.TYPESCRIPT, Language.JAVA),
                Arguments.of(Language.TYPESCRIPT, Language.TYPESCRIPT)
        );
    }

    @Disabled
    @Test
    void shouldCreateRepositoryWithJava() {
        // Act
        generator.generateNewRepository(repoName, Language.JAVA, Language.TYPESCRIPT);

        // Assert
        gitHub.verifyRepositoryExists();
        gitHub.verifyPathsExist(Language.JAVA, Language.TYPESCRIPT);
        gitHub.verifyDockerComposeImage(Language.JAVA, Language.TYPESCRIPT);
        gitHub.verifyReadmeHasBadges(Language.JAVA, Language.TYPESCRIPT);
        gitHub.verifyPagesEnabled();

        gitHub.verifyWorkflowsPass(Language.JAVA, Language.TYPESCRIPT);
    }

    @ParameterizedTest
    @MethodSource("languageProvider")
    void shouldCreateRepositoryWithLanguages(Language systemLanguage, Language systemTestLanguage) {
        generator.generateNewRepository(repoName, systemLanguage, systemTestLanguage);

        gitHub.verifyRepositoryExists();
        gitHub.verifyPathsExist(systemLanguage, systemTestLanguage);
        gitHub.verifyDockerComposeImage(systemLanguage, systemTestLanguage);
        gitHub.verifyReadmeHasBadges(systemLanguage, systemTestLanguage);
        gitHub.verifyPagesEnabled();

        gitHub.verifyWorkflowsPass(systemLanguage, systemTestLanguage);
    }

    @Test
    void shouldReturnErrorForInvalidLanguage() {
        generator.generateNewRepositoryExpectError(repoName, Language.NONE, Language.TYPESCRIPT);
    }

    private static String newName() {
        var repoName = "repo-" + System.currentTimeMillis();
        return repoName;
    }


}

File: console/tests/ScriptTest.cs
using Optivem.AtddAccelerator.TemplateGenerator.SystemTests.Clients;
using Optivem.AtddAccelerator.TemplateGenerator.SystemTests.Dsl;
using Optivem.AtddAccelerator.TemplateGenerator.SystemTests.Util;

namespace Optivem.AtddAccelerator.TemplateGenerator.SystemTests
{
    public class ScriptTest : IDisposable
    {
        private const string RepoOwner = "valentinajemuovic";

        private GeneratorDsl _generator;
        private GitHubDsl _gitHub;
        private string _repoName;

        public ScriptTest()
        {
            var generatorClient = new GeneratorClient();
            _generator = new GeneratorDsl(generatorClient);

            _repoName = NewName();
            var githubClient = new GithubClient(RepoOwner, _repoName);
            _gitHub = new GitHubDsl(githubClient);
        }

        public void Dispose()
        {
            var created = _generator.IsCreated(_repoName);

            if (created && _gitHub != null)
            {
                // _gitHub.DeleteRepository();
            }
        }

        public static IEnumerable<object[]> LanguageProvider()
        {
            return new List<object[]>
            {
                new object[] { Language.DotNet, Language.DotNet },
                new object[] { Language.DotNet, Language.Java },
                new object[] { Language.DotNet, Language.TypeScript },

                new object[] { Language.Java, Language.DotNet },
                new object[] { Language.Java, Language.Java },
                new object[] { Language.Java, Language.TypeScript },

                new object[] { Language.TypeScript, Language.DotNet },
                new object[] { Language.TypeScript, Language.Java },
                new object[] { Language.TypeScript, Language.TypeScript }
            };
        }

        [Theory]
        [MemberData(nameof(LanguageProvider))]
        public async Task ShouldCreateRepositoryWithLanguages(string systemLanguage, string systemTestLanguage)
        {
            await _generator.GenerateNewRepository(_repoName, systemLanguage, systemTestLanguage);

            _gitHub.VerifyRepositoryExists();
            _gitHub.VerifyPathsExist(systemLanguage, systemTestLanguage);
            _gitHub.VerifyDockerComposeImage(systemLanguage, systemTestLanguage);
            _gitHub.VerifyReadmeHasBadges(systemLanguage, systemTestLanguage);
            _gitHub.VerifyPagesEnabled();

            _gitHub.VerifyWorkflowsPass(systemLanguage, systemTestLanguage);
            // _gitHub.VerifyPackagesExist(systemLanguage);

            // TODO: VJ: Verify that only one package exists, rest are deleted

            // TODO: VJ: Rewrite like this
            // _gitHub.VerifyCommitStageSuccessful(systemLanguage); // -this checks only this path, the docker come, the readme badge, and that package was created
        }

        // TODO: Cannot create repository if name already exists
        // TODO: Error if gh not installed
        // TODO: Error if not logged into gh auth
        // TODO: Error if some mandatory parameter missing, or invalid, or empty
        // TODO: Error if no internet connection


        [Fact]
        public async Task QuickTest()
        {
            string systemLanguage = Language.Java;
            string systemTestLanguage = Language.TypeScript;

            await _generator.GenerateNewRepository(_repoName, systemLanguage, systemTestLanguage);

            _gitHub.VerifyRepositoryExists();
            _gitHub.VerifyPathsExist(systemLanguage, systemTestLanguage);
            _gitHub.VerifyDockerComposeImage(systemLanguage, systemTestLanguage);
            _gitHub.VerifyReadmeHasBadges(systemLanguage, systemTestLanguage);
            _gitHub.VerifyPagesEnabled();
        }

        [Fact]
        public async Task ShouldReturnErrorForInvalidSystemLanguage()
        {
            await _generator.GenerateNewRepositoryExpectError(_repoName, Language.None, Language.TypeScript);
        }

        private static string NewName()
        {
            var repoName = "repo-" + Guid.NewGuid().ToString();
            return repoName;
        }
    }
}